### DISCUSSION TITLE
Hardware Interrupts (zh-CN)

### DISCUSSION THREAD

**Discussion by giscus** [phil-opp/blog_os#1200] [Post Comments (translated)]
# Hardware Interrupts (zh-CN)

在本文中，我们会对可编程的中断控制器进行设置，以将硬件中断转发给CPU，而要处理这些中断，只需要像处理异常一样在中断描述符表中加入一个新条目即可，在这里我们会以获取周期计时器的中断和获取键盘输入为例进行讲解。

https://os.phil-opp.com/zh-CN/hardware-interrupts/

**Comment by shane-mason** [comment]
Wow! This is an amazing series. I wanted to build an OS to learn more about bare metal programming and after reading a bit, Rust was the obvious choice. The problem is, I've never programmed in Rust - I can't believe that after a weekend I have a basic system that I can start interacting with. Its bare bones and has a long way to be anything useful, but so far its been useful in teaching me Rust and bringing me back up to speed on x86 architecture. Well done.

**Comment by rasheedmhd** [comment]
The code below didn't work for me. I am don't really understand what the compile error message was saying. 

It sounded like, the time the blog was written, That was okay but as time passed by such code became hard errors.
```
lazy_static! {
        static ref KEYBOARD: Mutex<Keyboard<layouts::Us104Key, ScancodeSet1>> =
            Mutex::new(Keyboard::new(layouts::Us104Key, ScancodeSet1,
                HandleControl::Ignore)
            );
 }
```

I had to find away around it and i resorted to the implementation of the pc-keyboard trait and ended up with something that worked.


```
    lazy_static! {
        static ref KEYBOARD: Mutex<Keyboard<layouts::Us104Key, ScancodeSet1>> = {
            let keyboard = Mutex::new(Keyboard::new(layouts::Us104Key, ScancodeSet1, HandleControl::Ignore));
            keyboard
        };
    }
```

This is a link to my implementation, There might be something that skipped my mind.
https://github.com/Virux/oubre/blob/main/src/interrupts.rs

**Reply by bjorn3** [reply]
What was the error message?

**Reply by rasheedmhd** [reply]
I think I get it now. I tried reproducing the error and realized that it compiled on first try. I was confused but then I remember that I initially made a typo that I fixed later on. It appears to me that the error was coming from the typo.

The typo was that I passed `ScancodeSet` instead of `ScancodeSet1` to Keyboard. 

I am not sure this was the original error, but it contains a part of it so this is not a complete reproduce of the error.
```
   Compiling oubre_os v0.1.0 (/Users/Starlet/Dev/oubre_os)
error[E0412]: cannot find type `ScancodeSet` in this scope
   --> src/interrupts.rs:114:64
    |
114 |         static ref KEYBOARD: Mutex<Keyboard<layouts::Us104Key, ScancodeSet>> =
    |                                                                ^^^^^^^^^^^
    |
   ::: /Users/Starlet/.cargo/registry/src/github.com-1ecc6299db9ec823/pc-keyboard-0.5.1/src/scancodes.rs:8:1
    |
8   | pub struct ScancodeSet1;
    | ----------------------- similarly named struct `ScancodeSet1` defined here
    |
help: a struct with a similar name exists
    |
114 |         static ref KEYBOARD: Mutex<Keyboard<layouts::Us104Key, ScancodeSet1>> =
    |                                                                ~~~~~~~~~~~~
help: consider importing this trait
    |
1   | use pc_keyboard::ScancodeSet;
    |

error: impl method assumes more implied bounds than the corresponding trait method
   --> src/interrupts.rs:113:5
    |
113 | /     lazy_static! {
114 | |         static ref KEYBOARD: Mutex<Keyboard<layouts::Us104Key, ScancodeSet>> =
115 | |             Mutex::new(
116 | |                 Keyboard::new(layouts::Us104Key, ScancodeSet1, HandleControl::Ignore)
117 | |             );
118 | |     }
    | |_____^
    |
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #105572 <https://github.com/rust-lang/rust/issues/105572>
    = note: `#[deny(implied_bounds_entailment)]` on by default
    = note: this error originates in the macro `__lazy_static_internal` which comes from the expansion of the macro `lazy_static` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no method named `lock` found for struct `KEYBOARD` in the current scope
   --> src/interrupts.rs:127:33
    |
113 | /     lazy_static! {
114 | |         static ref KEYBOARD: Mutex<Keyboard<layouts::Us104Key, ScancodeSet>> =
115 | |             Mutex::new(
116 | |                 Keyboard::new(layouts::Us104Key, ScancodeSet1, HandleControl::Ignore)
117 | |             );
118 | |     }
    | |_____- method `lock` not found for this struct
...
127 |       let mut keyboard = KEYBOARD.lock();
    |                                   ^^^^ method not found in `KEYBOARD`

Some errors have detailed explanations: E0412, E0599.
For more information about an error, try `rustc --explain E0412`.```

**Reply by bjorn3** [reply]
The first error looks helpful to me at least, the third one could be better (to be precise I think the highlighted part of the code could be more specific), but the second one is just terrible. I don't understand what it says either.

**Reply by rasheedmhd** [reply]
Ahh! jezz, I felt same, about the last error. It is the part that I am referring to when I say the above error message contains  the original error. In the original error if the first part was as is, I am pretty sure I would have seen and fixed it but I didn't see it. Or maybe I just didn't check well, I can't be too sure at this point. Overall I am glad that you replied, I nearly held back asking. Thank You.

**Comment by olosso** [comment]
Hi! This is a fantastic series, and has definitely filled holes in my education.

I have a question regarding this bit: "This is because the keyboard controller won’t send another interrupt until we have read the so-called scancode of the pressed key."

How does the keyboard controller know that the scancode has been read from the port? Is there some sort of table somewhere, that defines readers and writers of/to a port, and every time a port is read from, a signal is sent to the party that writes to the port? Or what kind of mechanism is going on here? Thanks!

**Reply by bjorn3** [reply]
As I understand it reading from an IO port will directly ask the device (in this case the keyboard controller) for the result without buffering anything inside the CPU. This means that the keyboard controller knows the exact moment the scancode of the pressed key was read, as that is the moment where it got a read request for the respective IO port.

**Reply by olosso** [reply]
Oh, ok. Since you used the word "request", that makes it kind of sound like a client/server relationship, where the kernel is the client and the keyboard controller is the server.

**Comment by dporr** [comment]
A few questions:
* In the InterruptIndex struct implementations, why do we use the form `usize::from(self.as_u8())` instead of simply calling `self as usize` ?


* In the post you mention:
> To avoid another deadlock, we disable interrupts for the test’s duration. Otherwise, the test might get interrupted while the writer is still locked.

My understanding is that if we disable the interrupts, both the data race and the deadlock are eliminated. Doing a separate fix for both problems looks redundant, is this for didactic reasons only? or am I missing something?

**Comment by over04** [comment]
在64位系统下，按此教程中x86_64::instructions::interrupts::enable()会没有任何效果，需要按照下面的办法进行设置
1. **设置掩码**
```rust
let mut pics = interrupts::PICS.lock();
pics.initialize();
pics.write_masks(0b1111_1110, 0xFF); // 如果开启键盘中断，使用(0b1111_1100, 0xFF)
```
此时中断会发生，但是时钟中断一次后会马上引发double fault
2. **设置SS**
在64位模式下 SS 虽然不参与寻址，但在 iretq 返回时仍会校验它引用的是一个有效的可写数据段。因为 SS=0x10指向的是 TSS，iretq 会触发 #GP，随后在异常处理过程中又出错（不保证这个解释正确）
要做的就是添加kernel_data_segment并设置
```rust
// src/gdt.rs

let data_selector = gdt.append(Descriptor::kernel_data_segment());
...

// src/gdt.rs

SS::set_reg(GDT.data_selector);
```
然后时钟中断后就可以恢复，不会引发double fault
